frb_codegen/src/commands.rs

```diff
@@ -274,3 +280,91 @@ pub fn build_runner(dart_root: &str) -> Result {
     }
     Ok(())
 }
+
+fn guess_context(dart_root: &str) -> anyhow::Result<Context> {
+    debug!("Guessing context the runner is run into");
+    let pubspec = PathBuf::from(dart_root).join("pubspec.lock");
+    if !pubspec.exists() {
+        return Err(anyhow::Error::msg(format!(
+            "missing pubspec.lock in {}",
+            dart_root
+        )));
+    }
+    let pubspec = std::fs::read_to_string(pubspec)
+        .map_err(|_| anyhow::Error::msg(format!("unable to read pubspec.lock in {}", dart_root)))?;
+    let pubspec: Pubspec = serde_yaml::from_str(&pubspec).map_err(|_| {
+        anyhow::Error::msg(format!("unable to parse pubspec.lock in {}", dart_root))
+    })?;
+    if pubspec.packages.contains_key("flutter") {
+        return Ok(Context::Flutter);
+    }
+    Ok(Context::Dart)
+}
+
+#[derive(Debug, PartialEq)]
+enum Context {
+    Dart,
+    Flutter,
+}
+
+impl Context {
+    fn as_run_command(&self) -> &'static str {
+        match self {
+            Context::Dart => "dart",
+            Context::Flutter => "flutter pub",
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct Pubspec {
+    pub packages: HashMap<String, serde_yaml::Value>,
+}
+
+#[cfg(test)]
+mod tests {
+    use std::path::PathBuf;
+
+    use crate::commands::{guess_context, Context};
+
+    #[test]
+    fn guess_dart_context() {
```

Comment:
For these tests: shall we extract a common helper like `fn guess_context_base(path: Path, expect_context: Context)`, such that no duplicate of code for these.

Reply:
Done in a0f97a19d839719db0b3e1e3e0d029c0d4fe6d20 except `path` parameter being `PathBuf` for simplicity.

Reply:
You might actually like [test-case](https://crates.io/crates/test-case) crate if your testing base grows with similar units  :)

Reply:
Also renamed every var 'context' into 'toolchain' for clarity with de2f7ea5d97c34c301ad3e9ee20c742d951f1422

Reply:
if `PathBuf` annoys you in the method signature, I can still change it to `&Path` and convert back and forth with `.as_path()` and `.to_path_buf()`. Just let me know it's a one minute change.

Reply:
this is what I have in mind:
![test-changes](https://user-images.githubusercontent.com/21016014/181872530-0eb9e75d-433d-4b49-bfbe-9b389e26f69d.png)

Reply:
No worries - both LGTM :)

Reply:
ok concise way in 1bed76d2630edb5f629ecf020c41ef73bea8cac7